File: Scripts/spectrograms.m
% 1_generate_spectrograms_FINAL.m
% REVISION: Added 'audioinfo' safety check to prevent M1 Hangs.

clc; clear; close all;

% --- PATHS ---
BASE_PATH = '/Users/ahero1/Downloads/SonicID';
DATASET_PATH = fullfile(BASE_PATH, 'UrbanSound8K');
OUTPUT_PATH = fullfile(BASE_PATH, 'data', 'spectrograms');
CSV_FILE = fullfile(DATASET_PATH, 'UrbanSound8K.csv');

% --- SETTINGS ---
TARGET_FS = 22050;        
DURATION = 4;             
NUM_SAMPLES = TARGET_FS * DURATION;

% Filter Design (One-time setup)
Nyquist = TARGET_FS / 2;
[b, a] = butter(6, [50, 10000] / Nyquist, 'bandpass');

if ~exist(OUTPUT_PATH, 'dir'); mkdir(OUTPUT_PATH); end

% --- MAIN LOOP ---
opts = detectImportOptions(CSV_FILE);
opts.SelectedVariableNames = {'slice_file_name', 'fold', 'classID', 'class'};
metadata = readtable(CSV_FILE, opts);

disp(['Processing ', num2str(height(metadata)), ' files...']);
h = waitbar(0, 'Generating Spectrograms...');

for i = 1:height(metadata)
    try
        filename = metadata.slice_file_name{i};
        fold = metadata.fold(i);
        className = metadata.class{i};
        
        % Check for hidden files
        if startsWith(filename, '._'); continue; end
        
        audioPath = fullfile(DATASET_PATH, ['fold', num2str(fold)], filename);
        
        if ~isfile(audioPath); continue; end

        % --- SAFETY CHECK (Prevents Hangs) ---
        % 'audioinfo' is fast. If this fails, we skip 'audioread' completely.
        try
            info = audioinfo(audioPath);
        catch
            continue; % Silently skip bad files
        end
        % -------------------------------------

        [y, fs] = audioread(audioPath);
        
        % Processing Pipeline
        if size(y, 2) > 1; y = mean(y, 2); end
        if fs ~= TARGET_FS; y = resample(y, TARGET_FS, fs); end
        
        if length(y) < NUM_SAMPLES
            y = [y; zeros(NUM_SAMPLES - length(y), 1)];
        else
            y = y(1:NUM_SAMPLES);
        end
        
        y = filtfilt(b, a, y);
        
        % Spectrogram
        S = melSpectrogram(y, TARGET_FS, ...
            'WindowLength', 1024, ...
            'OverlapLength', 512, ...
            'NumBands', 64, ...
            'Range', [50, 10000]);
        
        S_dB = 10 * log10(S + eps);
        S_norm = (S_dB - min(S_dB(:))) / (max(S_dB(:)) - min(S_dB(:)));
        
        % Save
        classFolder = fullfile(OUTPUT_PATH, className);
        if ~exist(classFolder, 'dir'); mkdir(classFolder); end
        
        imgName = strrep(filename, '.wav', '.png');
        savePath = fullfile(classFolder, imgName);
        imwrite(S_norm, savePath);
        
    catch ME
        % Skip
    end
    
    if mod(i, 100) == 0
        waitbar(i/height(metadata), h, sprintf('Processed %d / %d', i, height(metadata)));
    end
end
close(h);
disp('Done.');

File: Scripts/spectrograms.py
import os
import pandas as pd
import numpy as np
import librosa
import soundfile as sf
from tqdm import tqdm
import matplotlib.pyplot as plt

# --- CONFIGURATION ---
# Base path (Adjust if needed, using relative path for safety)
BASE_PATH = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
DATASET_PATH = os.path.join(BASE_PATH, 'UrbanSound8K')
OUTPUT_PATH = os.path.join(BASE_PATH, 'data', 'spectrograms')
CSV_FILE = os.path.join(DATASET_PATH, 'UrbanSound8K.csv')

# Audio Settings
TARGET_FS = 22050
DURATION = 4.0
SAMPLES = int(TARGET_FS * DURATION)

# Create Output Directory
if not os.path.exists(OUTPUT_PATH):
    os.makedirs(OUTPUT_PATH)

print(f"Loading Metadata from: {CSV_FILE}")
try:
    df = pd.read_csv(CSV_FILE)
except FileNotFoundError:
    print(f"ERROR: CSV not found at {CSV_FILE}")
    exit()

print(f"Found {len(df)} files. Starting processing...")

# --- MAIN LOOP ---
for index, row in tqdm(df.iterrows(), total=df.shape[0]):
    try:
        filename = row['slice_file_name']
        fold = row['fold']
        class_name = row['class']
        
        # Construct Path
        audio_path = os.path.join(DATASET_PATH, f"fold{fold}", filename)
        
        # Skip if file doesn't exist
        if not os.path.exists(audio_path):
            continue
            
        # 1. Load Audio (Librosa handles M1 codecs perfectly)
        y, sr = librosa.load(audio_path, sr=TARGET_FS, mono=True)
        
        # 2. Pad or Crop to 4 seconds
        if len(y) < SAMPLES:
            padding = SAMPLES - len(y)
            y = np.pad(y, (0, padding), mode='constant')
        else:
            y = y[:SAMPLES]
            
        # 3. Generate Mel-Spectrogram
        S = librosa.feature.melspectrogram(y=y, sr=sr, n_fft=1024, hop_length=512, n_mels=64, fmin=50, fmax=10000)
        S_dB = librosa.power_to_db(S, ref=np.max)
        
        # 4. Save as Image
        # Create Class Folder
        class_folder = os.path.join(OUTPUT_PATH, class_name)
        if not os.path.exists(class_folder):
            os.makedirs(class_folder)
            
        save_name = filename.replace('.wav', '.png')
        save_path = os.path.join(class_folder, save_name)
        
        # Save output without axes/borders (Pure Image)
        plt.figure(figsize=(1, 1), dpi=100) # Small size is fine for CNN
        plt.axis('off')
        plt.axes([0., 0., 1., 1.], frameon=False, xticks=[], yticks=[])
        plt.imshow(S_dB, aspect='auto', origin='lower', cmap='magma')
        plt.savefig(save_path, bbox_inches='tight', pad_inches=0)
        plt.close()

    except Exception as e:
        print(f"Error processing {filename}: {e}")

print("Processing Complete. Check data/spectrograms folder.")